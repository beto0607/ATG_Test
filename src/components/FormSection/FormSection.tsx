import React, { useState } from 'react';
import { Form, FormControlProps, Button } from 'react-bootstrap';
import styles from './form-section.module.scss';
import { connect } from "react-redux";
import { getGameSchedule } from "../../actions/actions";

export interface ConnectedFormProps {
    getGameSchedule: any;
};
export const ConnectedForm: React.FC<ConnectedFormProps> = ({ getGameSchedule }: ConnectedFormProps) => {
    const [text, setText] = useState('');
    const handleSumbit = (event: React.FormEvent<HTMLFormElement>) => {
        event.preventDefault();
        if (text.length) {
            getGameSchedule(text);
        }
    };
    return (
        <div className={styles['form-section']}>
            <Form onSubmit={handleSumbit} noValidate={true} action='#'>
                <Form.Group controlId="gameType">
                    <Form.Label>Search for a game</Form.Label>
                    <Form.Control
                        type="text"
                        placeholder="Enter game type"
                        onChange={(event: React.FormEvent<FormControlProps>) => { setText(event.currentTarget.value || '') }}
                    />
                </Form.Group>
                <Button variant="primary" type="submit" block={true} disabled={!text}>
                    Search
                </Button>
            </Form>

        </div>
    );
};

function mapDispatchToProps(dispatch: any) {
    return {
        getGameSchedule: (text: string) => dispatch(getGameSchedule(text))
    };
}

const FormSection = connect(null, mapDispatchToProps)(ConnectedForm);
export default FormSection;
