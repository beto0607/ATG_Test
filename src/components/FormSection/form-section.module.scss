.form-section{

    width: 100%;
    padding: 20px;
    display: flex;
    justify-content: space-around;
    form{
        width: 400px;
    }
    
}