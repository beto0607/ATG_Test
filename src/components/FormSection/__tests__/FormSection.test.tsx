import React from 'react';
import { shallow } from 'enzyme';
import { ConnectedForm, ConnectedFormProps } from '../FormSection';
import toJson from 'enzyme-to-json'
import { Form, Button } from 'react-bootstrap'

const inputValue: string = 'input_value';
const inputValueEvent: any = { currentTarget: { value: inputValue } };
const emptyString: string = '';
const emptyStringEvent: any = { currentTarget: { value: emptyString } };
describe('ConnectedForm', () => {
  it('renders without crashing given the required props', () => {
    const component = shallow((<ConnectedForm {...{ getGameSchedule: jest.fn((text: string) => text) }} />));
    expect(toJson(component)).toMatchSnapshot()
  });

  it('should call getGameSchedule on submit with inputValue', () => {
    const props: ConnectedFormProps = { getGameSchedule: jest.fn((text: string) => text) };
    const component = shallow((<ConnectedForm {...props} />));
    // Updates text
    component.find(Form.Control).simulate('change', inputValueEvent);
    // Submits form
    component.find(Form).simulate('submit', { preventDefault: jest.fn() });
    expect(props.getGameSchedule).toHaveBeenCalledWith(inputValue);
  });

  it("should NOT call getGameSchedule on submit when input it's empty", () => {
    const props: ConnectedFormProps = { getGameSchedule: jest.fn((text: string) => text) };
    const component = shallow((<ConnectedForm {...props} />));
    // Updates text
    component.find(Form.Control).simulate('change', emptyStringEvent);
    // component.find(Button).simulate('click');
    component.find(Form).simulate('submit', { preventDefault: jest.fn() });
    expect(props.getGameSchedule).not.toHaveBeenCalled();
  });

  it("Button should be disabled as default", () => {
    const component = shallow((<ConnectedForm {...{ getGameSchedule: jest.fn((text: string) => text) }} />));
    expect(component.find(Button).props().disabled).toEqual(true);
  });

  it('Button should be enabled when text is NOT an empty string', () => {
    const component = shallow((<ConnectedForm {...{ getGameSchedule: jest.fn((text: string) => text) }} />));
    // Updates text
    component.find(Form.Control).simulate('change', inputValueEvent);
    expect(component.find(Button).props().disabled).toEqual(false);
  });

  it("Button should be disabled when input value is an empty string", () => {
    const component = shallow((<ConnectedForm {...{ getGameSchedule: jest.fn((text: string) => text) }} />));
    // Updates text
    component.find(Form.Control).simulate('change', emptyStringEvent);
    expect(component.find(Button).props().disabled).toEqual(true);
  });
})
