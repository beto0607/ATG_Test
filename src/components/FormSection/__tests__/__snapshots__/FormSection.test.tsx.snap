// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectedForm renders without crashing given the required props 1`] = `
<div
  className="form-section"
>
  <Form
    action="#"
    inline={false}
    noValidate={true}
    onSubmit={[Function]}
  >
    <FormGroup
      controlId="gameType"
    >
      <FormLabel
        column={false}
        srOnly={false}
      >
        Search for a game
      </FormLabel>
      <FormControl
        onChange={[Function]}
        placeholder="Enter game type"
        type="text"
      />
    </FormGroup>
    <Button
      active={false}
      block={true}
      disabled={true}
      type="submit"
      variant="primary"
    >
      Search
    </Button>
  </Form>
</div>
`;
